enhancements/auto-pruning.md

```diff
@@ -0,0 +1,149 @@
+
+### Non-Goals
+
+* Allowing users to specify multiple policies per repository
+* Applying policies based on repo or tag name pattern matching
+
+## Proposal
+
+This proposal is structured in three parts: how policies are managed, how the policies are structured in the DB, and how the policies will be executed via the auto-pruning worker.
+
+### Managing Policies
+
+A user is able to configure auto-pruning within the repository settings. To start, there will be two auto-pruning options:
+* Prune by number of tags, provided the number of tags
+* Prune tags older than date, provided the number of days 
+
+For this implementation users will only be able to configure one policy per repository. To interface with this new UI there will a new set of CRUD endpoints for managing auto-pruning policies. The endpoints returning lists will be paginated.
+
+**Create policy**
+```
+POST /api/v1/repository/{organization}/{repository}/autoprunepolicy
+By date:
+{
+    pruneBy: "creation_date",
+    olderThan: "2w"  
+}
+By number of tags:
+{
+    pruneBy: "number_of_tags",
+    numTags: 100  
+}
+```
+
+**Update policy**
+```
+PATCH /api/v1/organization/{organization}/{repository}/autoprunepolicy/{policyid}
+{
+    fieldToBeUpdated: value,
+}
+```
+
+**List policies under a repository**
+```
+GET /api/v1/organization/{organization}/{repository}/autoprunepolicy
+```
+
+**List policies under an organization**
+```
+GET /api/v1/organization/{organization}/autoprunepolicy
+```
+
+**Get policy**
+```
+GET /api/v1/organization/{organization}/{repository}/autoprunepolicy/{policyid}
+```
+
+**Delete policy**
+```
+DELETE /api/v1/organization/{organization}/{repository}/autoprunepolicy/{policyid}
+```
+
+### Database structure
+
+A table called `autoprunepolicy` will be used to store the repository auto-prune policies.
+| Field | Datatype | Attributes | Description|
+| --- | ----------- | ----------- | ----------- |
+| uuid | character varying(255) | Unique, not null, indexed | Unique identifier for this policy |
+| organization_id | integer | FK, not null | Organization that the policy falls under |
+| repository_id | integer | FK, not null | Repository that the policy is applied to |
+| config | text | JSON, not null | Policy configuration |
+| last_ran_ms | bigint | nullable, indexed | Last time this policy was executed |
+| is_running | boolean | default false | Policy is currently being executed |
+
+### Auto-pruning worker
+
+The auto-pruning worker is an asyncronous job that executes the policy and performs the following steps:
+* Worker starts every 30s or to some configured number
+* Select a policy with the least `last_ran_ms` and `is_running`=`false`
+    * A policy with a null `last_ran_ms` will also be selected first as this indicates it was never ran. This prioritizes policies that haven't been ran recently.
+* Set the policy to `is_running` = `true`
+* Determine which pruning method to use based on `config.pruneBy`
```

Comment:
What if there are two policies defined?

Reply:
multiple policies are supported in the redesign